contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+topology-aware scheduling logic.
+
+```
+// Value is “true” or “false”.  Default is false.
+const AlphaVolumeTopologySchedulingAnnotation = “volume.alpha.kubernetes.io/topology-scheduling”
+```
+
+While this approach can let us introduce the new behavior gradually, it has a
+few downsides:
+* StorageClass will be required to use this new logic, even if dynamic
+provisioning is not used (in the case of local storage).
+* We have to maintain two different paths for volume binding.
+* We will be depending on the storage admin to correctly configure the
+StorageClasses.
+
+For those reasons, we prefer to switch the behavior for all unbound PVCs and
+clearly communicate these changes to the community before changing the behavior
+by default in GA.
+
+### Integrating binding with scheduling
+For the alpha phase, the focus is on static provisioning of PVs to support
+persistent local storage.
+
+TODO: flow chart
+
+The proposed new workflow for volume binding is:
+1. Admin statically creates PVs and/or StorageClasses.
+2. User creates unbound PVC and there are no prebound PVs for it.
+3. **NEW:** PVC binding and provisioning is delayed until a pod is created that
+references it.
+4. User creates a pod that uses the PVC.
+5. Pod starts to get processed by the scheduler.
+6. **NEW:** A new predicate function, called MatchUnboundPVCs, will look at all of
+a Pod’s unbound PVCs, and try to find matching PVs for that node based on the
+PV topology.  If there are no matching PVs, then it checks if dynamic
+provisioning is possible for that node.
+7. **NEW:** The scheduler continues to evaluate priorities.  A new priority
+function, called PrioritizeUnboundPVCs, will get the PV matches per PVC per
+node, and compute a priority score based on various factors.
+8. **NEW:** After evaluating all the existing predicates and priorities, the
+scheduler will pick a node, and call a new assume function, AssumePVCs,
+passing in the Node.  The assume function will check if any binding or
+provisioning operations need to be done.  If so, it will update the PV cache to
+mark the PVs with the chosen PVCs.
+9. **NEW:** If PVC binding or provisioning is required, we do NOT AssumePod.
+Instead, a new bind function, BindPVCs, will be called asynchronously, passing
+in the selected node.  The bind function will prebind the PV to the PVC, or
+trigger dynamic provisioning, and then wait until the PVCs are bound
+successfully or encounter failure.  Then, it always sends the Pod through the
+scheduler again for reasons explained later.
+10. When a Pod makes a successful scheduler pass once all PVCs are bound, the
+scheduler assumes and binds the Pod to a Node.
+11. Kubelet starts the Pod.
+
+This new workflow will have the scheduler handle unbound PVCs by choosing PVs
+and prebinding them to the PVCs.  The PV controller completes the binding
+transaction, handling it as a prebound PV scenario.
+
+Prebound PVCs and PVs will still immediately be bound by the PV controller.
```

Comment:
What you mean by this? if they were prebound - why do they need binding?

Reply:
> if the user set PVC.volumeName, then the PV controller will set PV.ClaimRef.

That should not be the case... a PVC user cannot target a specific PV by name.

Reply:
I just tried it and it works.  I created 2 PVs of different sizes, and then created a PVC that requests the size of the smaller PV, but specifies the name of the larger PV.  And the volume got bound to the specified PV name.  From my understanding, it's been this way since the beginning.

Reply:
> From my understanding, it's been this way since the beginning.

I don't think so. It looks like that was added in 1.3 in https://github.com/kubernetes/kubernetes/pull/24331/files#diff-3aaafc34911c234537e0076b5ed40b37R202. I would not expect a namespaced user to be able to target a particular PV.

Reply:
Regardless, it's been like this for many releases now.  If we want to change it to not allow prebinding PVCs, then it can be considered separately from this proposal.  This proposal is not changing the current prebinding behavior.

Reply:
@liggitt, I thought it was common practice - a Pod author can set `pod.spec.nodeName` to bypass scheduling. This is the same thing.

Reply:
> a Pod author can set pod.spec.nodeName to bypass scheduling

that is prevented in some clusters